### ISSUE TITLE
[DataGrid] Pressing Shift + space should select new row without having to unselect the other row first

### ISSUE THREAD

**Issue by hangryernest** [mui-org/material-ui-x#3265]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Currently non multi select keyboard row selection on a datagrid requires a user to unselect a row, via shift + space, before another row can be selected.

### Expected behavior 🤔

Row selection for non multi select keyboard selection should behave the same as mouse click where it moves and selects the new row after a user presses shift + space

https://codesandbox.io/s/datagridprodemo-material-demo-forked-lostt?file=/demo.js

### Steps to reproduce 🕹

https://codesandbox.io/s/datagridprodemo-material-demo-forked-lostt?file=/demo.js

Steps:

1. Select a row in the grid with your mouse
2. navigate table to another row via arrows (stay in first column where edit cell won't be enabled)
3. click shift + space - nothing happens
4. navigate via keyboard to selected row, and deselect row with shift + space
5. navigate table to another row via arrows (stay in first column where edit cell won't be enabled)
6. click shift + space, new row is now selected

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Browser: Chrome 95.0.4638.69 
  System:
    OS: Windows 10 10.0.22000
  Binaries:
    Node: 14.16.1 - C:\Program Files\nodejs\node.EXE
    Yarn: Not Found
    npm: 6.14.12 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: Not Found
    Edge: Spartan (44.22000.120.0), Chromium (96.0.1054.29)
  npmPackages:
    @emotion/react: 11.5.0 => 11.5.0
    @emotion/styled: 11.3.0 => 11.3.0
    @mui/core:  5.0.0-alpha.54
    @mui/icons-material: 5.1.0 => 5.1.0
    @mui/lab: 5.0.0-alpha.54 => 5.0.0-alpha.54
    @mui/material: 5.1.0 => 5.1.0
    @mui/private-theming:  5.1.0
    @mui/styled-engine:  5.1.0
    @mui/styles: 5.1.0 => 5.1.0
    @mui/system:  5.1.0
    @mui/types:  7.1.0
    @mui/utils:  5.1.1
    @mui/x-data-grid-pro: 5.0.1 => 5.0.1
    @mui/x-license-pro:  5.0.1
    @types/react:  17.0.1
    react: 17.0.2 => 17.0.2
    react-dom: 17.0.2 => 17.0.2
```
</details>

### Order ID 💳 (optional)
#3197

**Comment by alexfauquette** [comment]
Thanks for reporting this bug.

The problem, is that we have a method `selectRow(id, isSelected, resetSelection)` used by `handleRowClick` and  `handleCellKeyDown`. The first one has complex logic to deal with <kbd>Ctrl</kbd> for example. The second only listen to <kbd>Shift</kbd>+<kbd>Space</kbd>

With such a structure, it seems difficult to maintain a logic that behaves the same for mouse and keyboard. I think the event logic could be redesigned as follow: 
- dummy functions which redirect events to the associated logic (select, edit, sorting, ...)
- dummy functions to modify the state
- The core logic in a single function

![image](https://user-images.githubusercontent.com/45398769/143225332-1dac9fe9-1d08-4caa-bfa9-64d28e2671c7.png)

This solution should solve issue #3272

It will force to have similar logic with mouse and keyboard. For example, default select (click or <kbd>Shift</kbd> +<kbd>Space</kbd> reset the selection)  and add to selection will be <kbd>Ctrl</kbd> + click or <kbd>Ctrl</kbd> + <kbd>Shift</kbd> +<kbd>Space</kbd>. Not sure if it is a good or bad thing

Do you have an opinion @mui-org/x ?

**Comment by m4theushw** [comment]
Why not drop `useGridKeyboard` and handle everything inside `useGridSelection`? "Keyboard" is not a feature, but "selection" is.

**Comment by flaviendelangle** [comment]
I'm interested to see what it would look like without `useGridKeyboard`. I agree that this hook feels off compared to the others.

**Comment by alexfauquette** [comment]
I think it will be renamed `useGridNavigation`. 

After moving the selection keyboard events in #3275 it will only contain navigation and column headers actions which should be moved too
